src/arch/arm/32/kernel/vspace.c

```diff
@@ -1730,6 +1768,17 @@ createSafeMappingEntries_PDE
             }
         }
 
+        /* Check that we are not overwriting an existing mapping */
+        if (pde_ptr_get_pdeType(ret.pde_entries.base) == pde_pde_section) {
+            paddr_t pde_paddr = pde_pde_section_ptr_get_address(ret.pde_entries.base);
+            if (pde_paddr && frame_asid == asidInvalid) {
```

Comment:
Would this check also work as `if (pde_paddr && pde_paddr != base)`? Which is encoding: Either the current pde entry isn't mapped, or if it is mapped, the destination paddr isn't being changed. Additionally, would it be better to test the actual bit for a valid mapping instead of a 0 paddr which could be a valid mapping address?

Reply:
I thought about doing it like that initially since it doesn't require you to pass around the frame_asid but the behaviour would be a little different between riscv and the other archs. 

As an example, suppose you have frame cap A which refers to frame X of physical memory. You do a `seL4_ARCH_Page_Map(frame_cap_A, vaddr, ...)` so that the page `vaddr` lies in is backed by frame X with frame cap A. You then make a copy of frame cap A with seL4_CNode_Copy, which gives you frame cap B. This also points to frame X of phsyical memory but has asid = `asidInvalid` and mapped_address = `NULL`.

If we implemented it as `if (pde_paddr && pde_paddr != base)`, I believe doing `seL4_ARCH_Page_Map(frame_cap_B, vaddr, ...)` would not work on RISCV but would work on the other architectures. 

As for the second part of your question, yeah I suppose that might be better. I thought a paddr of 0 would be an invalid address for some reason.

Reply:
Right, I think that it might be easier to follow if instead of passing the frame_asid in and testing whether it is invalid or not, to instead pass in a flag indicating whether the mapping is new or an update.  If it is new, it should then fail if there's already a mapping in the page table slot.  If it's an update then the asid is guaranteed to match because of earlier checks.  But what about if the physical address doesn't match? This could happen if the page table was deleted and then recreated as this wouldn't invalidate any existing frame caps.

Reply:
Right, but I thought this scenario fell under the other kind of stale cap behavior that was documented and should be allowed. Or do you think that this is something that needs to be addressed as well?